Sources/SpeziHealthKitBulkExport/BulkHealthExporter.swift

```
@@ -0,0 +1,112 @@
+//
+// This source file is part of the Stanford Spezi open-source project
+//
+// SPDX-FileCopyrightText: 2025 Stanford University and the project authors (see CONTRIBUTORS.md)
+//
+// SPDX-License-Identifier: MIT
+//
+
+import Foundation
+import HealthKit
+import Observation
+import Spezi
+import SpeziFoundation
+import SpeziHealthKit
+import SpeziLocalStorage
+
+
+@Observable
+public final class BulkHealthExporter: Module, EnvironmentAccessible, @unchecked Sendable {
+    @ObservationIgnored @Application(\.logger) var logger
+    @ObservationIgnored @Dependency(HealthKit.self) private var healthKit
+    @ObservationIgnored @Dependency(LocalStorage.self) private var localStorage
+    
+    /// All export sessions currently known to the Bulk Exporter.
+    @MainActor public private(set) var sessions: [any BulkExportSessionProtocol] = []
+    
+    /// Create a new Bulk Health Exporter
+    nonisolated public init() {}
+}
+
+
+extension BulkHealthExporter {
+    /// An error that can occur when creating or deleting a ``BulkExportSession``.
+    public enum SessionError: Error {
+        /// The ``BulkHealthExporter`` was unable to obtain a matching session,
+        /// since there is already a registered session with the same identifier, but a different ``BatchProcessor``.
+        case conflictingSessionAlreadyExists
+        /// The ``BulkHealthExporter`` was unable to delete the session, since there is already a matching registered session.
+        case unableToDeleteRegisteredSession
+    }
+    
+    /// The result of obtaining a ``BulkExportSession`` through the ``BulkHealthExporter``.
+    public enum ObtainSessionResult<Processor: BatchProcessor> {
+        /// The ``BulkHealthExporter`` created a new session.
+        ///
+        /// This case encapsulates both the newly created session, as well as an `AsyncStream` for getting the individual batch processing results.
+        case newSession(BulkExportSession<Processor>, AsyncStream<Processor.Output>)
```

What about moving the async stream as a response to the start function instead of the creation of the `BulkExportSession`? Would make this return cleaner and could allow us to easily attach this to a start of the export? And could also then make the start function either return an async stream or non if the output type is `Void`?

---

`start()` is currently allowed to be called multiple times, with any calls when the session is already running being no-ops.

i feel like having `start()` return the stream would not be the best idea; with the current design of `start()` and `pause()` returning void, it's easy to allow eg a UI component to control the export session, without requiring it to also deal with the actual batch processing results (which is the resposibility of the part of the code that starts the export session). otherwise, every place in the app that pauses/restarts a session would need to implement handling the results, and the app as a whole would need to somehow coordinate all of these.

regarding the Void-Output-based overloading: we should be able do that, though IMO we should rather do it with the `BulkHealthExporter.session()` function instead of `Session.start()`

---

@PSchmiedmayer what's your opinion here? do you see any potential issues with making the `start()` caller responsible for handling the session's batch results?

as stated above, this would IMO make it difficult to allow easily integrating session control into the UI.

typically, the part of the codebase that defines and creates the export session would also be the part that wants to receive and handle the individual batch results produced by the transformer; i can't really see any benefit from making the UI have to deal with this instead.

---

II see two possible mechanisms here:
1. Output is `Void`, therefore start doesn't return a stream and there is no stream to keep track of.
2. Output is non-void in which case we will get the output and have to keep track of it.

In both cases, the progress and eventual errors can be observed using all the properties that a session exposes.

I think that it is fine if a UI keeps track of these elements, but a better way (and what we should probably demonstrate in some of the documentation) would be writing an own `Module` or view model that creates a session and then keeps track of the AsyncStream output. This would also be the desired setup if we would attach the stream to the `session` call.
But in general there might not be any harm if a View `.task` modifier or similar element would keep track of it if needed and just wants to use the output to update some UI.

If we expose an async stream a second start call should probably throw an error to indicate that the session is already running.

---

that's exactly how i ended up implementing it. agree regarding the `Module` / view model point.

---

Thank you!